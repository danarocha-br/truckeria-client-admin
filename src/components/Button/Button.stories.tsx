import { Story } from '@storybook/react/types-6-0';
import { FiPlus } from 'react-icons/fi';

import Button from '.';

export default {
  title: 'Button',
  component: Button,
  args: {
    label: 'Sign in',
    ariaLabel: 'View more',
    kind: 'primary',
    size: 'md',
    fullWidth: false,
    disabled: false,
    loading: false,
    onClick: () => 'clicked',
  },
  argTypes: {
    label: {
      table: {
        category: 'Text',
      },
    },
    ariaLabel: {
      table: {
        category: 'Text',
      },
    },
    onClick: {
      table: {
        category: 'Events',
      },
    },
    variant: {
      table: {
        category: 'Modifiers',
      },
    },
    kind: {
      table: {
        category: 'Modifiers',
      },
    },
    size: {
      table: {
        category: 'Modifiers',
      },
    },
    fullWidth: {
      table: {
        category: 'Modifiers',
      },
    },
    icon: {
      type: '',
      table: {
        category: 'Modifiers',
      },
    },
    disabled: {
      type: '',
      table: {
        category: 'Modifiers',
      },
    },
    loading: {
      type: '',
      table: {
        category: 'Modifiers',
      },
    },
  },
};

const Template: Story = (args) => (
  <div className="flex flex-col items-center justify-center space-y-8 h-full w-full">
    <Button {...args} />

    <Button {...args} color="success" />
  </div>
);

export const Default = Template.bind({});

export const WithIcon = Template.bind({});
WithIcon.args = {
  label: 'Add',
  icon: <FiPlus></FiPlus>,
};

export const IconOnly = Template.bind({});
IconOnly.args = {
  ariaLabel: 'Add',
  icon: <FiPlus></FiPlus>,
  variant: 'icon',
};
