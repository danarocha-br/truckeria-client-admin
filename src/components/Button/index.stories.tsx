import { Button, IconButton } from '@chakra-ui/react';
import { Story, Meta } from '@storybook/react/types-6-0';
import { FiEdit2 } from 'react-icons/fi';

export default {
  title: 'Components/Button',
  component: Button,
  argTypes: {
    size: {
      control: {
        type: 'radio',
        options: ['sm', 'md'],
      },
    },
    variant: {
      control: {
        type: 'radio',
        options: ['primary', 'transparent'],
      },
    },
  },
  args: {
    size: 'md',
    isActive: false,
    isDisabled: false,
    isFullWidth: false,
    isLoading: false,
    variant: 'primary',
  },
} as Meta;

const Template: Story = (args) => <Button {...args}>I am a Label</Button>;

export const Primary = Template.bind({});

export const IconButtonDefault: Story = (args) => (
  <IconButton
    {...args}
    aria-label="action-name"
    variant="transparent"
    icon={<FiEdit2 />}
  ></IconButton>
);
