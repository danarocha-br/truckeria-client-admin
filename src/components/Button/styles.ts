import styled, { css, DefaultTheme } from 'styled-components';
import tw from 'twin.macro';

import { ButtonProps } from './';

type ButtonStyledProps = {
  hasIcon: boolean;
} & Omit<ButtonProps, 'label'>;

const sizeModifier = {
  sm: (variant: ButtonProps) => css`
    ${tw`
        text-sm
        font-bold
        leading-4
        h-10
        rounded-xs
    `};

    ${variant === 'icon' ? tw`w-10` : tw`px-10`};

    svg {
      ${tw`
        h-6
        w-6
    `};
    }
  `,

  md: (variant: ButtonProps) => css`
    ${tw`
        text-base
        leading-5
        h-12
        rounded-xs
    `}
    ${variant === 'icon' ? tw`w-14` : tw`py-10 px-16`};

    svg {
      ${tw`
        h-7
        w-7
    `};
    }
  `,

  lg: (variant: ButtonProps) => css`
    ${tw`
        text-lg
        leading-6
        h-16
        rounded-sm
    `}
    ${variant === 'icon' ? tw`w-16` : tw`py-12 px-20`};

    svg {
      ${tw`
        h-8
        w-8
    `};
    }
  `,
};

const iconModifier = {
  hasIcon: () => css`
    svg {
      ${tw`
        mr-4
        rounded-md
        w-6
        h-6
        -ml-4
      `}
    }
  `,
};

const colorModifier = {
  primary: (theme: DefaultTheme) => css`
    background-color: ${theme.action.primary.default};
    ${tw`focus:ring-brand`}

    &:hover {
      background-color: ${theme.action.primary.hover};
    }
    &:focus {
      background-color: ${theme.action.primary.hover};
      box-shadow: 0 0 0 1px ${theme.background.default},
        0 0 0 calc(2px + 2px) ${theme.action.primary.disabled};
    }
    &:active {
      background-color: ${theme.action.primary.pressed};
    }

    &:disabled {
      background-color: ${theme.action.primary.disabled};
    }
  `,

  success: (theme: DefaultTheme) => css`
    background-color: ${theme.action.success.default};

    &:hover {
      background-color: ${theme.action.success.hover};
    }
    &:focus {
      background-color: ${theme.action.success.hover};
      box-shadow: 0 0 0 1px ${theme.background.default},
        0 0 0 calc(2px + 2px) ${theme.action.success.disabled};
    }
    &:active {
      background-color: ${theme.action.success.pressed};
    }

    &:disabled {
      background-color: ${theme.action.success.disabled};
    }
  `,
};

export const Button = styled.button<ButtonStyledProps>`
  ${({ theme, size, fullWidth, hasIcon, color, variant }) => css`
    ${tw`
        text-white
        font-semibold
        text-center
        flex
        items-center
        justify-center
        whitespace-nowrap
        w-auto
        relative
        transition-all
        duration-300

        focus:outline-none
        focus:ring-4
    `}
    ${fullWidth && tw`w-full`}

    ${hasIcon && iconModifier.hasIcon()}
    ${!!size && sizeModifier[size](variant)}
    ${!!color && colorModifier[color](theme)}

    svg {
      ${variant === 'icon' && tw`ml-0 mr-0`};
    }
  `}
`;
