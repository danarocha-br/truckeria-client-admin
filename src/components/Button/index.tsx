import * as React from 'react';

import { AiOutlineLoading3Quarters } from 'react-icons/ai';
import * as S from './styles';

export type ButtonProps = {
  label?: string;
  ariaLabel?: string;
  icon?: JSX.Element;
  size?: 'sm' | 'md' | 'lg';
  fullWidth?: boolean;
  loading?: boolean;
  disabled?: boolean;
  color?: 'primary' | 'success';
  variant?: 'default' | 'icon';
  onClick?: () => (event: React.MouseEvent<HTMLButtonElement>) => void;
};

const Button = ({
  label,
  ariaLabel,
  size = 'md',
  fullWidth = false,
  loading = false,
  disabled = false,
  icon,
  variant = 'default',
  color = 'primary',
  onClick,
  ...props
}: ButtonProps) => (
  <S.Button
    aria-label={ariaLabel ? ariaLabel : label}
    size={size}
    fullWidth={fullWidth}
    hasIcon={!!icon}
    color={color}
    variant={variant}
    onClick={onClick}
    loading={loading}
    disabled={loading || disabled}
    {...props}
  >
    {!!icon && icon}
    {variant !== 'icon' && label}
    {loading && (
      <AiOutlineLoading3Quarters
        color="white"
        className="ml-5 -mr-4 animate-spin"
      />
    )}
  </S.Button>
);

export default Button;
