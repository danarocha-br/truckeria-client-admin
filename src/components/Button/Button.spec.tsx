import { screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { FiPlus } from 'react-icons/fi';

import { renderWithTheme } from 'utils/tests/helpers';

import Button from '.';

describe('<Button />', () => {
  it('should render the medium Button size by default', () => {
    const { container } = renderWithTheme(<Button label="button label" />);

    expect(screen.getByRole('button', { name: /button label/i })).toHaveStyle({
      padding: '2.5rem 4rem 2.5rem 4rem',
      'font-size': '1.1rem',
      height: '3rem',
    });

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render the small Button size if prop is passed', () => {
    renderWithTheme(<Button label="button label" size="sm" />);

    expect(screen.getByRole('button', { name: /button label/i })).toHaveStyle({
      padding: '2px 2.5rem 3px 2.5rem',
      'font-size': '0.8rem',
      height: '2.5rem',
    });
  });
});

it('should render the large Button size if prop is passed', () => {
  renderWithTheme(<Button label="button label" size="lg" />);

  expect(screen.getByRole('button', { name: /button label/i })).toHaveStyle({
    padding: '3rem 5rem 3rem 5rem',
    'font-size': '1.25rem',
    height: '4rem',
  });
});

it('should have fullwidth if prop is passed', () => {
  renderWithTheme(<Button label="button label" fullWidth />);

  expect(screen.getByRole('button', { name: /button label/i })).toHaveStyle({
    width: '100%',
  });
});

it('should render an icon if prop is passed', () => {
  renderWithTheme(
    <Button label="button label" icon={<FiPlus data-testid="icon" />} />
  );

  expect(screen.getByText(/button label/i)).toBeInTheDocument();
  expect(screen.getByTestId(/icon/i)).toBeInTheDocument();
});

it('Is accessible by tab', () => {
  renderWithTheme(<Button label="button label" />);

  const button = screen.getByLabelText(/button label/i);
  expect(document.body).toHaveFocus();

  userEvent.tab();
  expect(button).toHaveFocus();
});
