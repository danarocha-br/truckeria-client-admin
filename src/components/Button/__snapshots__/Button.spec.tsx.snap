// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> should render the medium Button size by default 1`] = `
.c0 {
  --tw-text-opacity: 1;
  color: rgba(255,255,255,var(--tw-text-opacity));
  font-weight: 600;
  text-align: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  white-space: nowrap;
  width: auto;
  position: relative;
  -webkit-transition-property: all;
  transition-property: all;
  -webkit-transition-timing-function: cubic-bezier(0.4,0,0.2,1);
  transition-timing-function: cubic-bezier(0.4,0,0.2,1);
  -webkit-transition-duration: 300ms;
  transition-duration: 300ms;
  font-size: 1.1rem;
  line-height: 1.25rem;
  height: 3.5rem;
  border-radius: 1.2rem;
  padding-top: 2.5rem;
  padding-bottom: 2.5rem;
  padding-left: 4rem;
  padding-right: 4rem;
  background-color: #FF8D4D;
}

.c0:focus {
  outline: 2px solid transparent;
  outline-offset: 2px;
  --tw-ring-offset-shadow: var(--tw-ring-inset) 0 0 0 var(--tw-ring-offset-width) var(--tw-ring-offset-color);
  --tw-ring-shadow: var(--tw-ring-inset) 0 0 0 calc(4px + var(--tw-ring-offset-width)) var(--tw-ring-color);
  box-shadow: var(--tw-ring-offset-shadow),var(--tw-ring-shadow),var(--tw-shadow,0 0 #0000);
}

.c0 svg {
  height: 1.75rem;
  width: 1.75rem;
}

.c0:focus {
  --tw-ring-opacity: 1;
  --tw-ring-color: rgba(255,141,77,var(--tw-ring-opacity));
}

.c0:hover {
  background-color: #ff6c1a;
}

.c0:focus {
  background-color: #ff6c1a;
  box-shadow: 0 0 0 1px #FFFFFF,0 0 0 calc(2px + 2px) rgba(255,141,77,0.6);
}

.c0:active {
  background-color: #e65300;
}

.c0:disabled {
  background-color: rgba(255,141,77,0.6);
}

<button
  aria-label="button label"
  class="c0"
  color="primary"
>
  button label
</button>
`;
